src/main.cpp

```diff
@@ -4515,12 +4574,21 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
         }
 
         CBlockIndex *pindexLast = NULL;
+        std::vector<CBlockIndex *> vToFetch;
+        bool fCanDirectFetch = CanDirectFetch(chainparams.GetConsensus());
         BOOST_FOREACH(const CBlockHeader& header, headers) {
             CValidationState state;
             if (pindexLast != NULL && header.hashPrevBlock != pindexLast->GetBlockHash()) {
                 Misbehaving(pfrom->GetId(), 20);
                 return error("non-continuous headers sequence");
             }
+            BlockMap::iterator it = mapBlockIndex.find(header.GetHash());
+            if (it != mapBlockIndex.end()) {
```

Comment:
It might be better to delete this check.  pindexLast shouldn't be set to a block that returns false from AcceptBlockHeader even if we already have the header.